src/NHibernate/Type/CollectionType.cs

```diff
@@ -390,15 +408,17 @@ public virtual object InstantiateResult(object original)
 			// One thing to be careful of here is a "bare" original collection
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
-			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			if(clearTargetsDirtyFlag)
 			{
-				if (!originalPc.IsDirty)
-					resultPc.ClearDirty();
+				resultPc.ClearDirty();
 			}
 
-			return result;
+			return target;
+		}
+
+		protected virtual bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
+		{ 
+			return CollectionHelper.CollectionEquals<object>(original.Cast<object>().ToArray(), target.Cast<object>().ToArray());
```

Comment:
I was not anticipating this `Cast<object>` with array conversion, which results in a double enumeration then followed by CollectionEquals own enumeration.
Please use `Enumerable.SequenceEqual` instead of `CollectionHelper.CollectionEquals`. This would avoid the `ToArray`. (And at that point, neither original nor target are supposed to be null, `SequenceEqual` can be safely used here without rechecking null.)